src/tools/rbd_mirror/instance_watcher/Types.h

```diff
@@ -0,0 +1,163 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef RBD_MIRROR_INSTANCE_WATCHER_TYPES_H
+#define RBD_MIRROR_INSTANCE_WATCHER_TYPES_H
+
+#include <string>
+#include <set>
+#include <boost/variant.hpp>
+
+#include "include/buffer_fwd.h"
+#include "include/encoding.h"
+#include "include/int_types.h"
+
+namespace ceph { class Formatter; }
+
+namespace rbd {
+namespace mirror {
+namespace instance_watcher {
+
+struct ImagePeer {
+  std::string mirror_uuid;
+  std::string image_id;
+  int64_t pool_id;
+
+  inline bool operator<(const ImagePeer &rhs) const {
+    return mirror_uuid < rhs.mirror_uuid;
+  }
+
+  inline bool operator==(const ImagePeer &rhs) const {
+    return (mirror_uuid == rhs.mirror_uuid && image_id == rhs.image_id &&
+	    pool_id == rhs.pool_id);
+  }
+
+  void encode(bufferlist& bl) const;
+  void decode(bufferlist::iterator& it);
+  void dump(Formatter *f) const;
+};
+
+WRITE_CLASS_ENCODER(ImagePeer);
+
+typedef std::set<ImagePeer> ImagePeers;
+
+enum NotifyOp {
+  NOTIFY_OP_IMAGE_ACQUIRE  = 0,
+  NOTIFY_OP_IMAGE_ACQUIRED = 1,
+  NOTIFY_OP_IMAGE_RELEASE  = 2,
+  NOTIFY_OP_IMAGE_RELEASED = 3,
+};
+
+struct ImageAcquirePayload {
+  static const NotifyOp NOTIFY_OP = NOTIFY_OP_IMAGE_ACQUIRE;
+
```

Comment:
... additionally, if you add a unique request id, you could have a generic "async op complete" callback instead of requiring individual "image acquired" / "image released" messages (similar to ```ImageWatcher```)